base/reduce.jl

```diff
@@ -168,10 +168,12 @@ foldr(op, itr) = mapfoldr(identity, op, itr)
 ## reduce & mapreduce
 
 function mapreduce_impl(f, op, A::AbstractArray, ifirst::Integer, ilast::Integer, blksize::Int=pairwise_blocksize(f, op))
-    if ifirst + blksize > ilast
+    if ifirst == ilast
+        @inbounds return r_promote(op, f(A[ifirst]))
```

Comment:
Actually, there's some code duplication between `_mapreduce()` and `mapreduce_impl()`. To fix that, `_mapreduce()` may call `mapreduce_impl()` for all but empty ranges.